components/browser/engine-gecko-nightly/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngine.kt

```diff
@@ -441,6 +443,15 @@ class GeckoEngine(
         return requireNotNull(webPushHandler)
     }
 
+    /**
+     * See [Engine.registerActivityDelegate].
+     */
+    override fun registerActivityDelegate(
+        activityDelegate: ActivityDelegate
+    ) {
+        runtime.activityDelegate = GeckoActivityDelegate(activityDelegate)
+    }
+
```

Comment:
How are apps going to implement `ActivityDelegate`? Will this need a reference to an `Activity`? Is this something we can add to sample browser?

Reply:
> How are apps going to implement `ActivityDelegate`? Will this need a reference to an `Activity`? 

I imagined the lifecycle for the activity delegate to be managed through the feature that implements it, but now I see that I would also need to introduce an unregister on the engine to notify it that we no longer have a delegate when we destroy from an activity or de-attach from a fragment.

> Is this something we can add to sample browser?

We could add something like this code below to the sample browser but it would tell the engine that we support WebAuthn features but our sample app cannot complete the request (needs to be an official signed app in the play store that is approved by Google to talk to their auth client).

```kotlin
// BaseBrowserFragment.kt
fun onStart() {
  engine.registerActivityDelegate(activityDelegate)
}

fun onStop() {
  engine.unregisterActivityDelegate()
}
```

We do have Reference Browser approved, so we can test this out in there before landing it in Fenix.

Reply:
This still feels somewhat fragile: We have one engine, but potentially multiple activities / fragments displaying web content. How do we know which one to delegate this to? What to do in cases where we do not have an activity?

I wish the GV team would have added that to `GeckoView`, the thing that is actually bound to an Activity/Fragment.

Also: The API has somewhat unexpected side effects. The activity delegate pretends to be generic, but it seems to enable AuthN as a side effect? I do not understand how this is supposed to work as a generic delegate if there will be other reason for this delegate in the future.

Reply:
> We have one engine, but potentially multiple activities / fragments displaying web content. How do we know which one to delegate this to? What to do in cases where we do not have an activity?

It would probably be the activity that the active GeckoSession is attached to with that engine. When we do not have an activity, do you mean the activity is destroyed? I've added an `unregister` method as well to remove the delegate in that case. If we're on a fragment, we could forward the activity events to the fragment in a similar way to [`UserInteractionHandler`](https://github.com/mozilla-mobile/android-components/blob/master/components/support/base/src/main/java/mozilla/components/support/base/feature/UserInteractionHandler.kt).

> I wish the GV team would have added that to GeckoView, the thing that is actually bound to an Activity/Fragment.

That's fair, I can bring this up with them.

> Also: The API has somewhat unexpected side effects. The activity delegate pretends to be generic, but it seems to enable AuthN as a side effect? I do not understand how this is supposed to work as a generic delegate if there will be other reason for this delegate in the future.

Good point, filed https://bugzilla.mozilla.org/show_bug.cgi?id=1671988

Reply:
I opened a separate issue for the request to move the `ActivityDelegate` from the engine to the view: https://bugzilla.mozilla.org/show_bug.cgi?id=1672195